File: games/monster-prom/plugin/EventDb.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace MonsterPromHelper.Ingame;

public sealed class EventRecord
{
    public string Key = "";
    public string Name = "";
    public string Route = "";
    public string Type = "";
    public List<OptionRecord> Options = new List<OptionRecord>();
}

public sealed class OptionRecord
{
    public int Option;
    public string Stat = "";
    public string Hint = "";
}

public sealed class EventDb
{
    public static readonly string[] StatKeys = { "SMARTS", "BOLD", "CREATIVE", "CHARM", "FUN", "MONEY" };

    private readonly Dictionary<string, EventRecord> _events = new(StringComparer.OrdinalIgnoreCase);
    private readonly HashSet<string> _secretEvents = new(StringComparer.OrdinalIgnoreCase);
    private readonly Dictionary<string, string> _secretEventToRoute = new(StringComparer.OrdinalIgnoreCase);
    private readonly Dictionary<string, string> _secretEventToWikiTitle = new(StringComparer.OrdinalIgnoreCase);
    private readonly Dictionary<string, List<string>> _secretChains = new(StringComparer.OrdinalIgnoreCase);
    private readonly List<SecretDialogHint> _secretHints = new List<SecretDialogHint>();

    public int EventCount => _events.Count;
    public int SecretHintCount => _secretHints.Count;

    public void LoadFromFolder(string dataFolder)
    {
        _events.Clear();
        _secretEvents.Clear();
        _secretEventToRoute.Clear();
        _secretEventToWikiTitle.Clear();
        _secretChains.Clear();
        _secretHints.Clear();

        var eventsPath = IoUtil.Combine(dataFolder, "events_db.json");
        if (File.Exists(eventsPath))
            LoadEvents(File.ReadAllText(eventsPath));

        var secretPath = IoUtil.Combine(dataFolder, "secret_endings.json");
        if (File.Exists(secretPath))
            LoadSecrets(File.ReadAllText(secretPath));

        BuildSecretHintIndex();
    }

    public static string ResolveDataFolder(string pluginRoot)
    {
        var candidates = new[]
        {
            IoUtil.Combine(pluginRoot, "data"),
            IoUtil.Combine(IoUtil.Combine(pluginRoot, "MonsterPromHelper"), "data"),
        };
        foreach (var dir in candidates)
        {
            if (File.Exists(IoUtil.Combine(dir, "events_db.json")))
                return dir;
        }
        return candidates[0];
    }

    private void LoadEvents(string json)
    {
        foreach (Match block in Regex.Matches(json, "\"([^\"]+)\"\\s*:\\s*\\{"))
        {
            var key = block.Groups[1].Value;
            var start = block.Index;
            var slice = json.Substring(start, Math.Min(4000, json.Length - start));
            var rec = new EventRecord
            {
                Key = key,
                Name = MatchStr(slice, "name") ?? key,
                Route = MatchStr(slice, "route") ?? "",
                Type = MatchStr(slice, "type") ?? "",
            };

            foreach (Match opt in Regex.Matches(slice, "\\{\\s*\"option\"\\s*:\\s*(\\d+)"))
            {
                var optSlice = json.Substring(
                    start + opt.Index,
                    Math.Min(500, json.Length - start - opt.Index));
                rec.Options.Add(new OptionRecord
                {
                    Option = int.Parse(opt.Groups[1].Value),
                    Stat = MatchStr(optSlice, "stat") ?? "",
                    Hint = MatchStr(optSlice, "hint") ?? "",
                });
            }

            if (rec.Options.Count > 0)
            {
                _events[rec.Name] = rec;
                _events[rec.Key] = rec;
            }
        }
    }

    private void LoadSecrets(string json)
    {
        LoadSecretWikiChains(json);

        var allMatch = Regex.Match(json, "\"all\"\\s*:\\s*\\[([^\\]]+)\\]", RegexOptions.Singleline);
        if (allMatch.Success)
            AddSecretNamesFromBlock(allMatch.Groups[1].Value);

        var routeIdx = json.IndexOf("\"by_route\"");
        if (routeIdx < 0)
            return;

        var section = json.Substring(routeIdx, Math.Min(6000, json.Length - routeIdx));
        var allIdx = section.IndexOf("\"all\"");
        if (allIdx > 0)
            section = section.Substring(0, allIdx);

        foreach (Match rm in Regex.Matches(section, "\"([^\"]+)\"\\s*:\\s*\\[([^\\]]*)\\]", RegexOptions.Singleline))
        {
            var routeName = rm.Groups[1].Value;
            if (string.Equals(routeName, "by_route", StringComparison.OrdinalIgnoreCase))
                continue;

            var chain = new List<string>();
            foreach (Match em in Regex.Matches(rm.Groups[2].Value, "\"([^\"]+)\""))
            {
                var ev = em.Groups[1].Value;
                if (ev.Length == 0)
                    continue;
                chain.Add(ev);
                _secretEvents.Add(ev);
                _secretEventToRoute[ev] = routeName;
            }

            if (chain.Count > 0 && !_secretChains.ContainsKey(routeName))
                _secretChains[routeName] = chain;
        }
    }

    private void AddSecretNamesFromBlock(string block)
    {
        foreach (Match em in Regex.Matches(block, "\"([^\"]+)\""))
            _secretEvents.Add(em.Groups[1].Value);
    }

    private void LoadSecretWikiChains(string json)
    {
        var idx = json.IndexOf("\"chains\"");
        if (idx < 0)
            return;

        var section = json.Substring(idx, System.Math.Min(14000, json.Length - idx));
        foreach (Match block in Regex.Matches(section, "\\{[^{}]*\"wiki_title\"[^{}]*\\}", RegexOptions.Singleline))
        {
            var slice = block.Value;
            var wikiTitle = MatchStr(slice, "wiki_title");
            var character = MatchStr(slice, "character") ?? "";
            var eventsMatch = Regex.Match(slice, "\"events\"\\s*:\\s*\\[([^\\]]*)\\]", RegexOptions.Singleline);
            if (!MonoUtil.HasText(wikiTitle) || !eventsMatch.Success)
                continue;

            var chain = new List<string>();
            foreach (Match em in Regex.Matches(eventsMatch.Groups[1].Value, "\"([^\"]+)\""))
            {
                var ev = em.Groups[1].Value;
                if (ev.Length == 0)
                    continue;
                chain.Add(ev);
                _secretEvents.Add(ev);
                _secretEventToWikiTitle[ev] = wikiTitle!;
                if (MonoUtil.HasText(character))
                    _secretEventToRoute[ev] = character;
            }

            if (chain.Count > 0)
                _secretChains[wikiTitle!] = chain;
        }
    }

    public string? GetSecretWikiTitle(string? eventName)
    {
        if (!MonoUtil.HasText(eventName))
            return null;
        string title;
        if (_secretEventToWikiTitle.TryGetValue(eventName.Trim(), out title))
            return title;
        return null;
    }

    private static string? MatchStr(string slice, string key)
    {
        var m = Regex.Match(slice, $"\"{key}\"\\s*:\\s*\"([^\"]*)\"");
        return m.Success ? m.Groups[1].Value : null;
    }

    public EventRecord? Get(string? name)
    {
        if (MonoUtil.IsNullOrWhiteSpace(name)) return null;
        return _events.TryGetValue(name.Trim(), out var ev) ? ev : null;
    }

    public bool IsSecretEnding(string? eventName) =>
        MonoUtil.HasText(eventName) && _secretEvents.Contains(eventName);

    public SecretAlertMatch? TryDetectSecret(LiveGameState live)
    {
        if (!live.EventActive || live.InCafeteriaPhase)
            return null;

        if (!DialogContext.IsSchoolEventDialog(live))
            return null;

        var o1 = (live.Option1Text ?? "").Trim();
        var o2 = (live.Option2Text ?? "").Trim();
        if (o1.Length < 10 || o2.Length < 10)
            return null;

        if (MonoUtil.HasText(live.EventName) && IsSecretEnding(live.EventName))
        {
            var evByName = Get(live.EventName);
            if (evByName != null)
            {
                var nameScore = ScoreOptionsAgainstEvent(evByName, o1, o2);
                if (nameScore >= 18)
                    return BuildSecretAlert(evByName, o1, o2, "ev:" + evByName.Name);
            }

            return null;
        }

        return MatchSecretByTwoOptions(o1, o2);
    }

    private SecretAlertMatch? MatchSecretByTwoOptions(string option1, string option2)
    {
        var scored = new List<KeyValuePair<int, EventRecord>>();
        foreach (var ev in SecretEventsOnly())
        {
            if (ev.Options.Count < 2)
                continue;

            var score = ScoreOptionsAgainstEvent(ev, option1, option2);
            if (score < 16)
                continue;

            scored.Add(new KeyValuePair<int, EventRecord>(score, ev));
        }

        if (scored.Count == 0)
            return null;

        scored.Sort((a, b) => b.Key.CompareTo(a.Key));
        var bestScore = scored[0].Key;
        var best = scored[0].Value;
        var secondScore = scored.Count > 1 ? scored[1].Key : 0;

        if (bestScore < 20)
            return null;
        if (scored.Count > 1 && bestScore < secondScore + 6)
            return null;

        return BuildSecretAlert(best, option1, option2, "dialog:" + best.Name);
    }

    private static int ScoreOptionsAgainstEvent(EventRecord ev, string option1, string option2)
    {
        if (ev.Options.Count < 2)
            return 0;

        var h1 = ev.Options[0].Hint ?? "";
        var h2 = ev.Options[1].Hint ?? "";
        if (h1.Length < 6 || h2.Length < 6)
            return 0;

        var ab = HintLineScore(h1, option1) + HintLineScore(h2, option2);
        var ba = HintLineScore(h1, option2) + HintLineScore(h2, option1);
        var score = ab > ba ? ab : ba;
        return score;
    }

    private SecretAlertMatch BuildSecretAlert(EventRecord ev, string option1, string option2, string dedupePrefix)
    {
        var wikiTitle = GetSecretWikiTitle(ev.Name);
        var route = GetSecretRouteName(ev.Name) ?? ev.Route;
        var chainKey = MonoUtil.HasText(wikiTitle) ? wikiTitle : route;
        var step = GetSecretStepIndex(ev.Name, chainKey);
        var total = MonoUtil.HasText(chainKey) && _secretChains.ContainsKey(chainKey)
            ? _secretChains[chainKey].Count
            : 0;
        var key = dedupePrefix + "|" + ev.Name + "|" + FingerprintDialog(option1, option2);
        var preview = TruncateOption(option1) + "  |  " + TruncateOption(option2);
        return BuildAlert(ev.Name, route, wikiTitle, step, total, 0, preview, key);
    }

    private static string TruncateOption(string text)
    {
        if (string.IsNullOrEmpty(text))
            return "";
        return text.Length <= 36 ? text : text.Substring(0, 33) + "...";
    }

    private static string FingerprintDialog(string option1, string option2)
    {
        var a = (option1 ?? "").Trim().ToLowerInvariant();
        var b = (option2 ?? "").Trim().ToLowerInvariant();
        if (a.Length > 48)
            a = a.Substring(0, 48);
        if (b.Length > 48)
            b = b.Substring(0, 48);
        return a + "||" + b;
    }

    private List<EventRecord> SecretEventsOnly()
    {
        var list = new List<EventRecord>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var kv in _events)
        {
            var ev = kv.Value;
            if (!IsSecretEventRecord(ev))
                continue;
            if (!seen.Add(ev.Name))
                continue;
            list.Add(ev);
        }
        return list;
    }

    private static SecretAlertMatch BuildAlert(
        string eventName,
        string route,
        string? wikiTitle,
        int step,
        int totalSteps,
        int option,
        string text,
        string dedupeKey)
    {
        return new SecretAlertMatch
        {
            EventName = eventName,
            Route = route,
            WikiTitle = wikiTitle ?? "",
            ChainStep = step,
            ChainTotal = totalSteps,
            MatchedOption = option,
            MatchedText = text ?? "",
            DedupeKey = dedupeKey,
        };
    }

    private void BuildSecretHintIndex()
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var kv in _events)
        {
            var ev = kv.Value;
            if (!IsSecretEventRecord(ev))
                continue;

            var dedupeName = ev.Name;
            if (!seen.Add(dedupeName))
                continue;

            for (var i = 0; i < ev.Options.Count; i++)
            {
                var opt = ev.Options[i];
                if (!MonoUtil.HasText(opt.Hint))
                    continue;
                _secretHints.Add(new SecretDialogHint
                {
                    EventName = ev.Name,
                    Route = ev.Route,
                    Option = opt.Option,
                    Hint = opt.Hint,
                });
            }
        }
    }

    private bool IsSecretEventRecord(EventRecord ev)
    {
        if (IsSecretEnding(ev.Name))
            return true;
        return MonoUtil.HasText(ev.Type)
            && string.Equals(ev.Type.Trim(), "Secret Ending", StringComparison.OrdinalIgnoreCase);
    }

    public string? GetSecretRouteName(string? eventName)
    {
        if (!MonoUtil.HasText(eventName))
            return null;
        string route;
        return _secretEventToRoute.TryGetValue(eventName.Trim(), out route) ? route : null;
    }

    public string BuildSecretBanner(string? eventName, string? eventType)
    {
        var wiki = GetSecretWikiTitle(eventName);
        if (MonoUtil.HasText(wiki))
        {
            var step = GetSecretStepIndex(eventName, wiki);
            var total = _secretChains.ContainsKey(wiki) ? _secretChains[wiki].Count : 0;
            if (total > 0 && step > 0)
                return "SECRET: " + wiki + "  (" + step + "/" + total + ")";
            return "SECRET: " + wiki;
        }

        var route = GetSecretRouteName(eventName);
        if (route != null)
        {
            var step = GetSecretStepIndex(eventName, route);
            var total = _secretChains.ContainsKey(route) ? _secretChains[route].Count : 0;
            if (total > 0 && step > 0)
                return "SECRET-ROUTE: " + route + "  (" + step + "/" + total + ")";
            return "SECRET-ROUTE: " + route;
        }

        if (MonoUtil.HasText(eventName) && IsSecretEnding(eventName))
            return "SECRET-ENDING EVENT";

        if (MonoUtil.HasText(eventType)
            && eventType.IndexOf("secret", StringComparison.OrdinalIgnoreCase) >= 0)
            return "SECRET-ENDING (Event-Typ)";

        return "";
    }

    private int GetSecretStepIndex(string? eventName, string route)
    {
        if (!MonoUtil.HasText(eventName) || !_secretChains.ContainsKey(route))
            return 0;

        var chain = _secretChains[route];
        var name = eventName.Trim();
        for (var i = 0; i < chain.Count; i++)
        {
            if (string.Equals(chain[i], name, StringComparison.OrdinalIgnoreCase))
                return i + 1;
        }

        return 0;
    }

    public Recommendation Recommend(EventRecord ev, IDictionary<string, int> stats)
    {
        var rec = new Recommendation();
        if (ev.Options.Count == 0) return rec;

        OptionRecord? best = null;
        var bestValue = int.MinValue;
        foreach (var opt in ev.Options)
        {
            var val = stats.TryGetValue(opt.Stat, out var v) ? v : 0;
            var maxRival = 0;
            foreach (var o in ev.Options)
            {
                if (o.Option == opt.Option) continue;
                if (stats.TryGetValue(o.Stat, out var rv) && rv > maxRival)
                    maxRival = rv;
            }
            string verdict;
            if (val > maxRival) verdict = "success";
            else if (val == maxRival) verdict = "tie";
            else verdict = "fail";

            rec.Lines.Add(new OptionLine
            {
                Option = opt.Option,
                Stat = opt.Stat,
                Value = val,
                Hint = opt.Hint,
                Verdict = verdict,
            });

            if (val > bestValue)
            {
                bestValue = val;
                best = opt;
            }
        }

        if (best != null)
        {
            rec.BestOption = best.Option;
            rec.BestStat = best.Stat;
            rec.BestValue = bestValue;
            rec.BestHint = best.Hint;
            foreach (var line in rec.Lines)
            {
                if (line.Option == best.Option)
                {
                    rec.BestVerdict = line.Verdict;
                    break;
                }
            }
        }

        return rec;
    }

    public List<EventRecord> SearchDialog(string query, int limit)
    {
        var results = new List<EventRecord>();
        var ranked = RankDialog(query, limit);
        foreach (var pair in ranked)
            results.Add(pair.Value);
        return results;
    }

    public HintMatchResult MatchByTwoOptions(
        string dialogText,
        string option1Line,
        string option2Line,
        bool relaxed)
    {
        var result = new HintMatchResult();
        var hay = (dialogText ?? "").ToLowerInvariant();
        var a = (option1Line ?? "").Trim();
        var b = (option2Line ?? "").Trim();
        var hasLines = a.Length >= 3 && b.Length >= 3;

        var scored = new List<KeyValuePair<int, EventRecord>>();
        foreach (var ev in UniqueEvents())
        {
            if (ev.Options.Count < 2)
                continue;

            var h1 = ev.Options[0].Hint ?? "";
            var h2 = ev.Options[1].Hint ?? "";
            if (h1.Length == 0 || h2.Length == 0)
                continue;

            var minLine = relaxed ? 8 : 10;
            var minFull = relaxed ? 4 : 5;
            int score;
            if (hasLines)
            {
                var ab = HintLineScore(h1, a) + HintLineScore(h2, b);
                var ba = HintLineScore(h1, b) + HintLineScore(h2, a);
                score = ab > ba ? ab : ba;
                if (score < minLine)
                    continue;
            }
            else
            {
                var s1 = HintLineScore(h1, hay);
                var s2 = HintLineScore(h2, hay);
                if (s1 < minFull || s2 < minFull)
                    continue;
                score = s1 + s2;
            }

            scored.Add(new KeyValuePair<int, EventRecord>(score, ev));
        }

        scored.Sort((x, y) =>
        {
            var c = y.Key.CompareTo(x.Key);
            return c != 0 ? c : string.Compare(x.Value.Name, y.Value.Name, StringComparison.OrdinalIgnoreCase);
        });

        for (var i = 0; i < scored.Count && i < 12; i++)
            result.Hits.Add(scored[i].Value);

        if (result.Hits.Count == 0)
        {
            result.Status = hasLines
                ? "No event matches the on-screen choices"
                : "No event — waiting for both choice buttons";
            return result;
        }

        var bestScore = scored[0].Key;
        var best = scored[0].Value;
        var secondScore = scored.Count > 1 ? scored[1].Key : 0;
        var pickScore = relaxed ? 11 : 14;
        var gap = relaxed ? 3 : 4;
        var soloScore = relaxed ? 14 : 18;

        if (bestScore >= pickScore && (scored.Count == 1 || bestScore >= secondScore + gap))
        {
            result.Best = best;
            result.Status = "ok";
            return result;
        }

        if (bestScore >= soloScore)
        {
            result.Best = best;
            result.Status = "ok";
            return result;
        }

        result.Status = "Multiple events possible — check the list";
        return result;
    }

    /// <summary>No yield — iterator state machines need CurrentManagedThreadId (.NET 4+).</summary>
    private List<EventRecord> UniqueEvents()
    {
        var list = new List<EventRecord>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var kv in _events)
        {
            if (!seen.Add(kv.Value.Name))
                continue;
            list.Add(kv.Value);
        }
        return list;
    }

    private List<KeyValuePair<int, EventRecord>> RankDialog(string query, int limit)
    {
        var results = new List<KeyValuePair<int, EventRecord>>();
        var q = (query ?? "").Trim().ToLowerInvariant();
        if (q.Length < 2)
            return results;

        var words = SplitWords(q);
        if (words.Count == 0)
            words.Add(q);

        foreach (var ev in UniqueEvents())
        {
            var name = ev.Name.ToLowerInvariant();
            var key = ev.Key.ToLowerInvariant();
            var route = (ev.Route ?? "").ToLowerInvariant();
            var etype = (ev.Type ?? "").ToLowerInvariant();
            var hints = new System.Text.StringBuilder();
            foreach (var opt in ev.Options)
                hints.Append((opt.Hint ?? "").ToLowerInvariant()).Append(' ');

            var haystack = name + " " + key + " " + route + " " + etype + " " + hints;

            var score = 0;
            if (q.Contains(name) || name.Contains(q) || q.Contains(key))
                score += 8;
            foreach (var word in words)
            {
                if (word.Length < 3)
                    continue;
                if (haystack.Contains(word))
                    score += 3;
                if (hints.ToString().Contains(word))
                    score += 4;
                if (route.Contains(word))
                    score += 2;
            }

            if (score > 0)
                results.Add(new KeyValuePair<int, EventRecord>(score, ev));
        }

        results.Sort((a, b) => b.Key.CompareTo(a.Key));
        if (results.Count > limit)
            results = results.GetRange(0, limit);
        return results;
    }

    private static List<string> SplitWords(string q)
    {
        var words = new List<string>();
        var part = new System.Text.StringBuilder();
        foreach (var ch in q)
        {
            if (char.IsLetterOrDigit(ch))
                part.Append(ch);
            else if (part.Length > 0)
            {
                if (part.Length >= 3)
                    words.Add(part.ToString());
                part.Length = 0;
            }
        }
        if (part.Length >= 3)
            words.Add(part.ToString());
        return words;
    }

    private static int HintLineScore(string hint, string line)
    {
        if (hint.Length == 0 || line.Length == 0)
            return 0;

        var h = hint.ToLowerInvariant().Trim();
        var ln = line.ToLowerInvariant().Trim();
        var score = 0;
        if (ln.Contains(h) || h.Contains(ln))
            score += 12;

        var hi = 0;
        var match = 0;
        var maxLen = h.Length < ln.Length ? h.Length : ln.Length;
        while (hi < h.Length && match < ln.Length && maxLen > 0)
        {
            if (h[hi] == ln[match])
            {
                score += 1;
                hi++;
            }
            match++;
        }

        foreach (var word in SplitWords(h))
        {
            if (ln.Contains(word))
                score += 3;
        }

        return score;
    }
}

public sealed class HintMatchResult
{
    public EventRecord? Best;
    public List<EventRecord> Hits = new List<EventRecord>();
    public string Status = "";
}

public sealed class Recommendation
{
    public List<OptionLine> Lines = new List<OptionLine>();
    public int BestOption;
    public string BestStat = "";
    public int BestValue;
    public string BestHint = "";
    public string BestVerdict = "";
}

public sealed class OptionLine
{
    public int Option;
    public string Stat = "";
    public int Value;
    public string Hint = "";
    public string Verdict = "";
}

public sealed class SecretDialogHint
{
    public string EventName = "";
    public string Route = "";
    public int Option;
    public string Hint = "";
}

File: python/src/event_db.py
"""Load and query Monster Prom event database."""

from __future__ import annotations

import json
import re
from difflib import SequenceMatcher
from pathlib import Path


# Same order as in-game stat sheet: Smarts, Boldness, Creativity, Charm, Fun, Money
STAT_KEYS = ("SMARTS", "BOLD", "CREATIVE", "CHARM", "FUN", "MONEY")


class EventDatabase:
    def __init__(self, path: Path) -> None:
        self.path = path
        self.events: dict[str, dict] = {}
        if path.exists():
            self.events = json.loads(path.read_text(encoding="utf-8"))

    def get(self, name: str) -> dict | None:
        return self.events.get(name.lower()) or self.events.get(name)

    def search(self, query: str, limit: int = 25) -> list[dict]:
        q = query.strip().lower()
        if not q:
            return []
        results = []
        for key, ev in self.events.items():
            name = ev.get("name", key)
            if q in key or q in name.lower():
                results.append(ev)
                if len(results) >= limit:
                    break
        return results

    def _rank_dialog(self, query: str, limit: int = 10) -> list[tuple[int, dict]]:
        q = query.strip().lower()
        if len(q) < 2:
            return []

        words = [w for w in re.split(r"[^a-z0-9]+", q) if len(w) >= 3]
        if not words:
            words = [q]

        scored: list[tuple[int, dict]] = []
        for key, ev in self.events.items():
            name = ev.get("name", key)
            route = (ev.get("route") or "").lower()
            etype = (ev.get("type") or "").lower()
            hints = " ".join(
                (o.get("hint") or "").lower() for o in ev.get("options", [])
            )
            haystack = f"{name.lower()} {key} {route} {etype} {hints}"

            score = 0
            if q in name.lower() or q in key:
                score += 8
            for word in words:
                if word in haystack:
                    score += 3
                if word in hints:
                    score += 4
                if word in route:
                    score += 2

            if score > 0:
                scored.append((score, ev))

        scored.sort(key=lambda item: (-item[0], len(item[1].get("name", ""))))
        return scored[:limit]

    def search_dialog(self, query: str, limit: int = 10) -> list[dict]:
        """Search by words from visible dialog — hints, route, LI names, event name."""
        return [ev for _, ev in self._rank_dialog(query, limit)]

    def best_dialog_match(
        self, query: str, limit: int = 8
    ) -> tuple[dict | None, list[dict]]:
        """Pick one event when OCR/text match is strong enough to auto-apply."""
        ranked = self._rank_dialog(query, limit)
        if not ranked:
            return None, []

        best_score, best = ranked[0]
        if best_score < 8:
            return None, [ev for _, ev in ranked]

        if len(ranked) == 1:
            return best, [ev for _, ev in ranked]

        second_score = ranked[1][0]
        if best_score >= second_score + 5 and best_score >= 11:
            return best, [ev for _, ev in ranked]
        if best_score >= 16:
            return best, [ev for _, ev in ranked]
        return None, [ev for _, ev in ranked]

    def _hint_tokens(self, hint: str) -> list[str]:
        return [w for w in re.split(r"[^a-z0-9]+", hint.lower()) if len(w) >= 3]

    def _hint_line_score(self, hint: str, line: str) -> int:
        if not hint or not line:
            return 0
        h = hint.lower().strip()
        ln = line.lower().strip()
        score = 0
        if h in ln or ln in h:
            score += 12
        ratio = SequenceMatcher(None, h, ln).ratio()
        score += int(ratio * 10)
        for word in self._hint_tokens(hint):
            if word in ln:
                score += 3
        return score

    def match_screenshot(
        self,
        ocr_text: str,
        option1_line: str = "",
        option2_line: str = "",
    ) -> tuple[dict | None, list[dict], str, str]:
        """
        Match event from screenshot OCR: option lines first, then full-text fallbacks.
        Returns (best, hits, status, detail_message).
        """
        best, hits, status = self.match_by_two_options(
            ocr_text, option1_line, option2_line, relaxed=True
        )
        if best and status == "ok":
            return best, hits, status, ""

        combined = " ".join(
            part
            for part in (option1_line, option2_line, ocr_text)
            if part and part.strip()
        ).strip()
        if combined:
            best, hits = self.best_dialog_match(combined, limit=10)
            if best:
                return best, hits, "ok", "Full-text search"
            if hits:
                return None, hits, "multiple", "Multiple hits — check the list"
            hits = self.search_dialog(combined, limit=10)
            if len(hits) == 1:
                return hits[0], hits, "ok", "Keyword search"
            if hits:
                return None, hits, "multiple", "Keyword — check the list"

        return None, [], "no_hit", "OCR text matches no event in the DB"

    def match_by_two_options(
        self,
        ocr_text: str,
        option1_line: str = "",
        option2_line: str = "",
        limit: int = 12,
        *,
        relaxed: bool = False,
    ) -> tuple[dict | None, list[dict], str]:
        """
        Find event where both answer hints match OCR (best: two option lines).
        Returns (best_event, ranked_hits, status_message).
        """
        hay = (ocr_text or "").lower()
        a = (option1_line or "").strip()
        b = (option2_line or "").strip()
        has_lines = len(a) >= 3 and len(b) >= 3

        scored: list[tuple[int, dict]] = []
        for key, ev in self.events.items():
            opts = ev.get("options") or []
            if len(opts) < 2:
                continue
            h1 = (opts[0].get("hint") or "").strip()
            h2 = (opts[1].get("hint") or "").strip()
            if not h1 or not h2:
                continue

            min_line = 8 if relaxed else 10
            min_full = 4 if relaxed else 5
            if has_lines:
                assign_ab = self._hint_line_score(h1, a) + self._hint_line_score(h2, b)
                assign_ba = self._hint_line_score(h1, b) + self._hint_line_score(h2, a)
                score = max(assign_ab, assign_ba)
                if score < min_line:
                    continue
            else:
                s1 = self._hint_line_score(h1, hay)
                s2 = self._hint_line_score(h2, hay)
                if s1 < min_full or s2 < min_full:
                    continue
                score = s1 + s2

            scored.append((score, ev))

        scored.sort(key=lambda item: (-item[0], item[1].get("name", "")))
        hits = [ev for _, ev in scored[:limit]]
        if not hits:
            msg = "No event matches the recognized choices"
            if has_lines:
                msg += f" — read: „{a[:40]}“ / „{b[:40]}“"
            return None, [], msg

        best_score, best = scored[0]
        second = scored[1][0] if len(scored) > 1 else 0
        pick_score = 11 if relaxed else 14
        gap = 3 if relaxed else 4
        solo_score = 14 if relaxed else 18

        if best_score >= pick_score and (len(scored) == 1 or best_score >= second + gap):
            return best, hits, "ok"

        if best_score >= solo_score:
            return best, hits, "ok"

        return None, hits, "multiple"

    def all_names(self) -> list[str]:
        return sorted(ev.get("name", k) for k, ev in self.events.items())


def recommend_option(event: dict, stats: dict[str, int]) -> list[dict]:
    """Return options annotated with success likelihood vs other option stats."""
    options = event.get("options", [])
    if not options:
        return []

    annotated = []
    for opt in options:
        stat = opt.get("stat", "")
        value = stats.get(stat, 0)
        other_stats = [o["stat"] for o in options if o["stat"] != stat]
        rival_values = [stats.get(s, 0) for s in set(other_stats)]
        max_rival = max(rival_values) if rival_values else 0
        if value > max_rival:
            verdict = "success"
        elif value == max_rival:
            verdict = "tie"
        else:
            verdict = "fail"

        annotated.append(
            {
                **opt,
                "value": value,
                "verdict": verdict,
            }
        )

    best = max(annotated, key=lambda o: o["value"])
    for o in annotated:
        o["recommended"] = o["option"] == best["option"] and o["verdict"] == "success"

    return annotated


def best_option(annotated: list[dict]) -> dict | None:
    winners = [o for o in annotated if o.get("recommended")]
    if winners:
        return winners[0]
    if annotated:
        return max(annotated, key=lambda o: o.get("value", 0))
    return None
